### PR TITLE
[Python] Removed mutable default parameters

### PR DESCRIPTION
**Pull request by Tishj**
This PR addresses the 'Mutable Objects' portion of issue #4342 

By changing every parameter that takes a mutable object as parameter, where we supplied a default value, this should be fixed now.

The code however did not get prettier from this though.. Everywhere that we accept a `connection` object, we now default to `None` and then assign the default_connection inside the function when the connection is None.

I also came across [this](https://pybind11.readthedocs.io/en/stable/advanced/smart_ptrs.html#std-shared-ptr) post that describes how you shouldn't return a raw pointer of an object that's wrapped in pybind as being a shared_ptr.

Because python/pybind will wrap this raw pointer in a shared_ptr and thus the object gets freed twice.

It sounds like this issue applies to `DuckDBPyConnection::DefaultConnection()`, but since we have yet to experience any issues from this, I'm not sure if I'm missing something.
EDIT: this does not apply to `default_connection` because we store a reference to the object in the `duckdb` module as an attribute, so the object never goes out of scope (we always keep this one reference alive)

### CODE REVIEW

File: tools/pythonpkg/src/pyconnection.cpp [hunk 1]
```diff
@@ -600,14 +615,21 @@ static unique_ptr<TableFunctionRef> ScanReplacement(ClientContext &context, cons
 }
 
 shared_ptr<DuckDBPyConnection> DuckDBPyConnection::Connect(const string &database, bool read_only,
-                                                           const py::dict &config_dict) {
+                                                           py::object config_options) {
```
Comment (Mause) [inline tools/pythonpkg/src/pyconnection.cpp:618]: This should remain of type `py::dict`

File: tools/pythonpkg/src/pyconnection.cpp [hunk 2]
```diff
@@ -283,6 +289,12 @@ unique_ptr<DuckDBPyRelation> DuckDBPyConnection::Values(py::object params) {
 	if (!connection) {
 		throw ConnectionException("Connection has already been closed");
 	}
+	if (params.is_none()) {
+		params = py::list();
+	}
+	if (!py::hasattr(params, "__len__")) {
+		throw InvalidInputException("Type of object passed to parameter 'values' has to be <list>");
```
Comment (Mause) [inline tools/pythonpkg/src/pyconnection.cpp:296]:
```suggestion
		throw InvalidInputException("Type of object passed to parameter 'values' must be iterable");
```

**Review by Mause — CHANGES_REQUESTED** [review]

File: tools/pythonpkg/src/pyconnection.cpp (hunk 1 above)
Comment (Tishj) [reply tools/pythonpkg/src/pyconnection.cpp:618]: py::dict is not a fan of my passing py::none() as default to it
I thought we wanted to remove all mutable objects as default argument, or did I miss interpret the issue?

Comment (Tishj) [reply tools/pythonpkg/src/pyconnection.cpp:618]:
```py
tools/pythonpkg/tests/fast/test_insert.py:9: in TestInsert
    conn = duckdb.connect()
E   TypeError: connect(): incompatible function arguments. The following argument types are supported:
E       1. (database: str = ':memory:', read_only: bool = False, config: dict = None) -> duckdb.DuckDBPyConnection
E   
E   Invoked with:
```

Comment (Mause) [reply tools/pythonpkg/src/pyconnection.cpp:618]: Yes, that error message is the result (to my understanding) of us not using `std::optional`

We should be using the `py::dict` as the type of the arg, it's the value of the default that should not be mutable

Comment (Mause) [reply tools/pythonpkg/src/pyconnection.cpp:618]: Yes, that is what I mentioned in our meeting a couple of weeks ago. I don't think adding a single file header library is a huge thing?

**Pull request merged**